agent/consul/state/catalog.go

```diff
@@ -642,6 +671,12 @@ func (s *Store) ensureServiceTxn(tx *memdb.Txn, idx uint64, node string, svc *st
 	if err := tx.Insert("index", &IndexEntry{"services", idx}); err != nil {
 		return fmt.Errorf("failed updating index: %s", err)
 	}
+	if !hasSameTags {
+		// We need to update /catalog/services only tags are different
+		if err := tx.Insert("index", &IndexEntry{"service-catalog", idx}); err != nil {
```

Comment:
This PR includes a functionality change for the index update optimization from your other PR into a "fix" for an unrelated bug.

It would be much easier to reason about just fixing this one bug in isolation without any change to behaviour in this PR and discuss the issues with the actual change of behaviour in a separate PR.

Reply:
True... but in both cases, it will have a conflict... I was thinking that https://github.com/hashicorp/consul/pull/3934#discussion_r173527678 was about to be merged , so I took it as a base... but it can wait most probably

Reply:
Hello @banks ,

As requested, this PR: https://github.com/hashicorp/consul/pull/3965 is the same, but without depending of /catalog/services optimization